### PR TITLE
Hybrid Pruning

### PR DESCRIPTION
**Pull request by varunagrawal**
Added pruning of the eliminated hybrid bayes net.

### CODE REVIEW

File: gtsam/hybrid/HybridBayesNet.cpp [hunk 1]
```diff
@@ -14,3 +14,96 @@
...
+/// Return the DiscreteKey vector as a set.
+static std::set<DiscreteKey> DiscreteKeysAsSet(const DiscreteKeys &dkeys) {
+  std::set<DiscreteKey> s;
+  s.insert(dkeys.begin(), dkeys.end());
+  return s;
+}
+
+/* ************************************************************************* */
+HybridBayesNet HybridBayesNet::prune(
+    const DecisionTreeFactor::shared_ptr &discreteFactor) const {
```
Comment (ProfFan) [inline gtsam/hybrid/HybridBayesNet.cpp:30]: nit: I would argue that we should only allow pruning through a GaussianMixture

Comment (varunagrawal) [reply gtsam/hybrid/HybridBayesNet.cpp:30]: I'm not sure I follow. Isn't pruning done on the discrete decision tree and then reflected on the full hybrid tree?

A GaussianMixture would only have the tree on it's keys right? Not the full graph

File: gtsam/hybrid/tests/testHybridIncremental.cpp [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+//       CHECK(lastDensity(assignment));
+//       EXPECT(assert_equal(*unprunedLastDensity(assignment),
+//                           *lastDensity(assignment)));
+//     }
+//   }
+// }
+
+/* ****************************************************************************/
+// // Test approximate inference with an additional pruning step.
+// TEST(DCGaussianElimination, Incremental_approximate) {
```
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridIncremental.cpp:349]: still commented out, need to merge other before merging this

File: gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+  EXPECT_LONGS_EQUAL(4, hybridBayesNet->size());
+  EXPECT_LONGS_EQUAL(1, remainingFactorGraph->size());
+}
+
+/* ************************************************************************* */
+int main() {
+  TestResult tr;
+  return TestRegistry::runAllTests(tr);
+}
+/* ************************************************************************* */
```
Comment (ProfFan) [inline gtsam/hybrid/tests/testHybridNonlinearFactorGraph.cpp:761]: newline

**Review by ProfFan — APPROVED** [review]
Most of the code here are reviewed in other PRs

**Pull request merged**